flow/model/qvalue/avro_converter.go

```diff
@@ -679,3 +680,220 @@ func (ns *NumericStat) CollectWarnings(warnings *shared.QRepWarnings) {
 		*warnings = append(*warnings, warning)
 	}
 }
+
+// ComputeAvroSize calculates the estimated uncompressed Avro-encoded size for a QValue
+// based on the schema determined by GetAvroSchemaFromQValueKind
+// Note: currently the size computation assumes ClickHouse as the TargetDWH
+func ComputeAvroSize(value types.QValue, nullable bool, logger log.Logger) int64 {
+	// we only use union tag for nullable types
+	unionTagSize := int64(0)
+	if nullable {
+		unionTagSize = 1
+	}
+
+	if value.Value() == nil {
+		return unionTagSize
+	}
+
+	// Avro encodes arrays with a zero-byte terminator
+	arrTerminationByteSize := int64(1)
+
+	size := int64(0)
+	switch v := value.(type) {
```

Comment:
Seems a little bad that the switch branches are duplicated when the size is tied to the value. How about:
```go
func QValueToAvro(..., calcSize bool) (any, int64, error) {
    switch v := value.(type) {
    case types.QValueFloat64:
        return c.processNullableUnion(float64(v.Val)), constSize(8, calcSize), nil
    case types.QValueString:
        return c.processNullableUnion(v.Value()), stringSize(v.Value(), calcSize), nil
    ...
}

func constSize(n int64, calcSize bool) int64 {
    if !calcSize {
        return 0
    }
    return n
}

func stringSize(s string, calcSize bool) int64 {
    if !calcSize {
        return 0
    }
    return varIntSize(int64(len(s))) + int64(len(s))
}
```

Reply:
i actually started with this, and then switched to the current way for readability (since size can be 0), but the switch branches being a bit wasteful is a good point.